File: eco-core/core-nms/v1_16_R2/src/main/java/com/willfp/eco/proxy/v1_16_R2/VillagerTrade.java
package com.willfp.eco.proxy.v1_16_R2;

import com.willfp.eco.core.display.Display;
import com.willfp.eco.proxy.proxies.VillagerTradeProxy;
import org.bukkit.craftbukkit.v1_16_R2.inventory.CraftItemStack;
import org.bukkit.craftbukkit.v1_16_R2.inventory.CraftMerchantRecipe;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.MerchantRecipe;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Field;

public final class VillagerTrade implements VillagerTradeProxy {
    @Override
    public void displayTrade(@NotNull final MerchantRecipe merchantRecipe) {
        try {
            // Bukkit MerchantRecipe result
            Field fResult = MerchantRecipe.class.getDeclaredField("result");
            fResult.setAccessible(true);
            ItemStack result = merchantRecipe.getResult().clone();
            Display.display(result);
            fResult.set(merchantRecipe, result);

            // Get NMS MerchantRecipe from CraftMerchantRecipe
            Field fHandle = CraftMerchantRecipe.class.getDeclaredField("handle");
            fHandle.setAccessible(true);
            net.minecraft.server.v1_16_R2.MerchantRecipe handle = (net.minecraft.server.v1_16_R2.MerchantRecipe) fHandle.get(merchantRecipe); // NMS RecipeR

            Field fSelling = net.minecraft.server.v1_16_R2.MerchantRecipe.class.getDeclaredField("sellingItem");
            fSelling.setAccessible(true);

            ItemStack selling = CraftItemStack.asBukkitCopy(handle.sellingItem).clone();
            Display.display(selling);

            fSelling.set(handle, CraftItemStack.asNMSCopy(selling));
        } catch (IllegalAccessException | NoSuchFieldException e) {
            e.printStackTrace();
        }
    }
}

File: eco-core/core-nms/v1_19_R2/src/main/kotlin/com/willfp/eco/internal/spigot/proxy/v1_19_R2/VillagerTrade.kt
package com.willfp.eco.internal.spigot.proxy.v1_19_R2

import com.willfp.eco.core.display.Display
import com.willfp.eco.internal.spigot.proxy.VillagerTradeProxy
import net.minecraft.nbt.CompoundTag
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.trading.MerchantOffer
import org.bukkit.craftbukkit.v1_19_R2.inventory.CraftItemStack
import org.bukkit.craftbukkit.v1_19_R2.inventory.CraftMerchantRecipe
import org.bukkit.entity.Player
import org.bukkit.inventory.MerchantRecipe
import java.lang.reflect.Field

class VillagerTrade : VillagerTradeProxy {
    private val handle: Field = CraftMerchantRecipe::class.java.getDeclaredField("handle")

    override fun displayTrade(
        recipe: MerchantRecipe,
        player: Player
    ): MerchantRecipe {
        recipe as CraftMerchantRecipe

        val nbt = getHandle(recipe).createTag()
        for (tag in arrayOf("buy", "buyB", "sell")) {
            val nms = ItemStack.of(nbt.getCompound(tag))
            val displayed = Display.display(CraftItemStack.asBukkitCopy(nms), player)
            val itemNBT = CraftItemStack.asNMSCopy(displayed).save(CompoundTag())
            nbt.put(tag, itemNBT)
        }

        return CraftMerchantRecipe(MerchantOffer(nbt))
    }

    private fun getHandle(recipe: CraftMerchantRecipe): MerchantOffer {
        return handle[recipe] as MerchantOffer
    }

    init {
        handle.isAccessible = true
    }
}